Refactor minimal .ipynb model into more traditional vivarium model form

aflaxman opened the pull request:
## Title: Refactor code from notebook into repo
<!-- Ideally, <=50 chars. 50 chars is here..: -->

### Description
<!-- For use in commit message, wrap at 72 chars. 72 chars is here: -->
- *Category*: <!-- one of bugfix, data artifact, implementation, observers,
                   post-processing, refactor, revert, test, release, other/misc -->refactor
- *JIRA issue*: none
- *Research reference*: <!--Link to research documentation for code --> https://gist.github.com/aflaxman/4165105a6922b6f373546484a6dd7dc3


### Changes and notes
<!-- 
Change description – why, what, anything unexplained by the above.
Include guidance to reviewers if changes are complex.
--> I expected it be straightforward, but I was getting an error that I have preserved below.  After a round of fixes, it is working now.

### Verification and Testing
<!--
Details on how code was verified. Consider: plots, images, (small) csv files.
-->
In my first attempt, `simulate run` produced a cryptic error
```
simulate run -v --pdb src/vivarium_sodium_reduction/model_specifications/model_spec.yaml
...
  File "/homes/abie/.conda/envs/vivarium_nih_us_cvd/lib/python3.11/site-packages/vivarium/framework/population/population_view.py", line 175, in get
    raise PopulationError(
vivarium.framework.population.exceptions.PopulationError: Requested column(s) {'high_systolic_blood_pressure_exposure'} not in population table. This is likely due to a failure to require some columns, randomness streams, or pipelines when registering a simulant initializer, a value producer, or a value modifier. NOTE: It is possible for a run to succeed even if resource requirements were not properly specified in the simulant initializers or pipeline creation/modification calls. This success depends on component initialization order which may change in different run settings.
```
after fixes discussed in this PR, it is now able to run successfully:
```
simulate run -v --pdb src/vivarium_sodium_reduction/model_specifications/model_spec.yaml
...
2024-08-19 14:53:27.531 | vivarium.framework.engine:258 - Event: time_step__cleanup
2024-08-19 14:53:27.535 | vivarium.framework.engine:264 - Updating: 10000
2024-08-19 14:53:27.574 | vivarium.framework.engine:258 - Event: collect_metrics
2024-08-19 14:53:27.578 | vivarium.framework.engine:264 - Updating: 10000
2024-08-19 14:53:27.607 | vivarium.framework.engine:306 - 
              Event  Count  Mean time (s)  Std. dev. time (s)  Total time (s)  % Total time
     initialization      1           0.14                0.00            0.14          0.06
              setup      1         112.15                0.00          112.15         47.25
         post_setup      1           0.00                0.00            0.00          0.00
population_creation      1           1.69                0.00            1.69          0.71
 time_step__prepare     44           2.10                0.05           92.37         38.92
          time_step     44           0.64                0.02           28.35         11.94
 time_step__cleanup     44           0.04                0.00            1.87          0.79
    collect_metrics     44           0.02                0.00            0.79          0.33
     simulation_end      1           0.00                0.00            0.00          0.00
              total      1         237.37                0.00          237.37        100.00
2024-08-19 14:53:27.613 | vivarium.interface.cli:158 - Simulation finished.
```

aflaxman committed:
WIP: refactor .ipynb model into .py files

albrja commented on setup.py:
```diff
@@ -44,8 +44,8 @@
-        "vivarium==2.3.8",
-        "vivarium_public_health==2.3.3",
+        "vivarium==3.0.0",
+        "vivarium_public_health==3.0.0",
```
You will want to set these to >=3.0.0, <4.0.0

albrja commented on src/vivarium_sodium_reduction/components/risks.py:
```diff
@@ -0,0 +1,168 @@
+
+        return post_processor
+
+class CorrelatedRisk(DropValueRisk):
```
Could you add a docstring here to explain what this class is doing? The risk methods I assume you are inheriting are abstracted away at this point and the level at which you have subclassed here it is hard to understand what exactly you are getting/trying to do with this class.

aflaxman replied:
TBH, I just asked an AI to adapt it from the https://github.com/ihmeuw/vivarium_nih_us_cvd model and this is what I got.  I'll add a stub of a docstring, and perhaps @hussain-jafari can suggest something more informative.

albrja reviewed (approved):
Approving but would like two minor updates to be made.

hussain-jafari commented on src/vivarium_sodium_reduction/components/risks.py:
```diff
@@ -0,0 +1,168 @@
+
+import numpy as np
+import pandas as pd
+import scipy
```
add `from gbd_mapping import risk_factors` under `import scipy`

hussain-jafari commented on src/vivarium_sodium_reduction/components/risks.py:
```diff
@@ -0,0 +1,168 @@
+
+    @property
+    def initialization_requirements(self) -> Dict[str, List[str]]:
+        return {"requires_columns": ["age"]}
```
Line 93 should be updated to `return {"requires_columns": ["age"] + self.ensemble_propensities}`

hussain-jafari commented on src/vivarium_sodium_reduction/components/risks.py:
```diff
@@ -0,0 +1,168 @@
+        self.correlated_risks = [EntityString(risk) for risk in correlated_risks]
+        self.correlation_matrix = correlation_matrix
+        self.propensity_column_names = [f"{risk.name}_propensity" for risk in self.correlated_risks]
+        self.exposure_column_names = [f"{risk.name}_exposure" for risk in self.correlated_risks]
```
We need to add an attribute as follows:

```
self.ensemble_propensities = [
    f"ensemble_propensity_" + risk
    for risk in self.correlated_risks
    if risk_factors[risk.name].distribution == "ensemble"
]
```

hussain-jafari commented on src/vivarium_sodium_reduction/model_specifications/model_spec.yaml:
```diff
@@ -3,37 +3,62 @@ components:
+                                'cause.ischemic_heart_disease.incidence_rate')
+    #     - NonLogLinearRiskEffect('risk_factor.high_systolic_blood_pressure',
+    #                             'cause.ischemic_stroke.incidence_rate')
+    vivarium_sodium_reduction.components:
```
This component needs to be moved into the vivarium_sodium_reduction.components block above because we unfortunately can't have two keys with the same name in our model spec.

aflaxman replied:
Oh, annoying. It would be nice if that were allowed, or at least had a clearer error message to indicate that it is not allowed.  Can you add a JIRA ticket to the backlog about it for me, @hussain-jafari ?

hussain-jafari replied:
Done! Ticket at https://jira.ihme.washington.edu/browse/MIC-5256

aflaxman committed:
ENH: fixes to the refactor to make it work

Thanks to @hussain-jafari for the detailed fixes; also includes tweaks from Jim's review.

hussain-jafari commented:
@aflaxman I think this PR is ready to merge (after linting it looks like). I can update the risk component docstrings in a separate PR
